ed25519.go

```diff
@@ -0,0 +1,173 @@
+
+var (
+	onceSupportsEd25519 sync.Once
+	supportsEd25519     bool
+)
+
+// SupportsEd25519 returns true if the current OpenSSL version supports
+// GenerateKeyEd25519, NewKeyFromSeedEd25519, SignEd25519 and VerifyEd25519.
+func SupportsEd25519() bool {
+	onceSupportsEd25519.Do(func() {
+		switch vMajor {
+		case 1:
+			supportsEd25519 = version1_1_1_or_above()
+		case 3:
+			name := C.CString("ED25519")
+			defer C.free(unsafe.Pointer(name))
+			sig := C.go_openssl_EVP_SIGNATURE_fetch(nil, name, nil)
+			if sig != nil {
+				C.go_openssl_EVP_SIGNATURE_free(sig)
+				supportsEd25519 = true
+			}
+		}
+	})
+	return supportsEd25519
+}
+
+// GenerateKeyEd25519 generates a public/private key pair.
+func GenerateKeyEd25519() (pub, priv []byte, err error) {
+	pkey, err := generateEVPPKey(C.GO_EVP_PKEY_ED25519, 0, "")
+	if err != nil {
+		return nil, nil, err
+	}
+	defer C.go_openssl_EVP_PKEY_free(pkey)
+	priv = make([]byte, privateKeySizeEd25519)
+	err = extractPKEYPrivEd25519(pkey, priv)
+	if err != nil {
+		return nil, nil, err
+	}
+	pub = make([]byte, publicKeySizeEd25519)
+	copy(pub, priv[seedSizeEd25519:])
+	return pub, priv, nil
+}
+
+// NewKeyFromSeedEd25519 calculates a private key from a seed. It will panic if
+// len(seed) is not [SeedSize]. This function is provided for interoperability
+// with RFC 8032. RFC 8032's private keys correspond to seeds in this
+// package.
+func NewKeyFromSeedEd25519(seed []byte) (priv []byte, err error) {
+	// Outline the function body so that the returned key can be stack-allocated.
+	priv = make([]byte, privateKeySizeEd25519)
+	err = newKeyFromSeedEd25519(priv, seed)
+	if err != nil {
+		return nil, err
+	}
+	return priv, err
+}
+
+func newKeyFromSeedEd25519(priv []byte, seed []byte) error {
+	if len(seed) != seedSizeEd25519 {
+		panic("ed25519: bad seed length: " + strconv.Itoa(len(seed)))
+	}
+	pkey := C.go_openssl_EVP_PKEY_new_raw_private_key(C.GO_EVP_PKEY_ED25519, nil, base(seed), C.size_t(len(seed)))
+	if pkey == nil {
+		return newOpenSSLError("EVP_PKEY_new_raw_private_key")
+	}
+	defer C.go_openssl_EVP_PKEY_free(pkey)
+	return extractPKEYPrivEd25519(pkey, priv)
+}
+
+func extractPKEYPrivEd25519(pkey C.GO_EVP_PKEY_PTR, priv []byte) error {
+	pubSize, privSize := C.size_t(publicKeySizeEd25519), C.size_t(seedSizeEd25519)
+	if C.go_openssl_EVP_PKEY_get_raw_public_key(pkey, base(priv[seedSizeEd25519:]), &pubSize) != 1 {
+		return newOpenSSLError("EVP_PKEY_get_raw_public_key")
+	}
+	if pubSize != publicKeySizeEd25519 {
+		panic("ed25519: bad public key length: " + strconv.Itoa(int(pubSize)))
+	}
+	if C.go_openssl_EVP_PKEY_get_raw_private_key(pkey, base(priv), &privSize) != 1 {
+		return newOpenSSLError("EVP_PKEY_get_raw_private_key")
+	}
+	if privSize != seedSizeEd25519 {
+		panic("ed25519: bad private key length: " + strconv.Itoa(int(privSize)))
+	}
+	return nil
+}
+
+// SignEd25519 signs the message with priv and returns a signature.
+func SignEd25519(priv, message []byte) (sig []byte, err error) {
```

Comment:
Unlike ECDSA, this API takes a private key as bytes and internally recovers the corresponding EVP_PKEY; that I suspect would prevent using non-extractable keys such as on PKCS#11. Why not defining a dedicated key types around EVP_PKEY?

Reply:
Good point. I've updated the implementation to use `PrivateKeyEd25519` and `PublicKeyEd25519` instead of raw byte slices.